enterprise/server/scheduling/task_router/task_router.go

```diff
@@ -124,35 +124,41 @@ func (tr *taskRouter) RankNodes(ctx context.Context, cmd *repb.Command, remoteIn
 		return nonePreferred(nodes)
 	}
 
-	nodeByID := map[string]interfaces.ExecutionNode{}
+	nodeByExecutorID := map[string]interfaces.ExecutionNode{}
+	nodeByHostID := map[string]interfaces.ExecutionNode{}
```

Comment:
Doesn't this assume there's only one executor per host? I think this should be a multi-map (or map to a slice) and then the logic below should probably select a random executor if there are multiple on a preferred host. Mind adding a test for that case too?

Reply:
I don't think random routing is the right behavior, because we don't support multiple executors per host unless they are using different filecache directories, in which case picking a random executor might route the task to the wrong one (the filecache might not have the task's inputs).

Ideally, I think the behavior would be "route to the executor that last executed the task, or if it doesn't exist, route to the executor that now lives on that host." Implementing this behavior would complicate how we store data in Redis, though. Since we don't really recommend or support running multiple executors per host (except in unit tests, where we fake the host ID to force different filecache dirs), I would prefer to keep this as-is for now.

Reply:
Yeah, I think the behaviour you describe is what we'd like, but I suppose it's not necessary at the moment. Do you think you could add a comment in the code elaborating?

Reply:
Done